go/pkg/services/logcollector/server.go

```diff
@@ -477,71 +490,47 @@ func (s *Server) startLogStreaming(ctx context.Context,
+	logFile *os.File,
 	stream io.ReadCloser) (bool, error) {
 
-	// create a reader from the stream, to allow peeking into it
-	streamReader := bufio.NewReader(stream)
-
-	// wait for the stream to have logs before reading them
-	if !s.hasLogs(ctx, runUID, streamReader) {
-		s.Logger.WarnWithCtx(ctx, "Stream doesn't have logs or context has been canceled", "runUID", runUID)
-		return false, nil
-	}
-
-	// open log file in read/write and append, to allow reading the logs while we write more logs to it
-	openFlags := os.O_RDWR | os.O_APPEND
-	file, err := os.OpenFile(logFilePath, openFlags, 0644)
-	if err != nil {
-		s.Logger.WarnWithCtx(ctx, "Failed to open file", "err", err, "logFilePath", logFilePath)
-		return true, errors.Wrapf(err, "Failed to open file in path %s", logFilePath)
-	}
-	defer file.Close() // nolint: errcheck
-
-	// spin a goroutine that will unblock `CopyBuffer` if context is dead
-	copyBufferDone := make(chan struct{}, 1)
-	go func() {
-		select {
-		case <-ctx.Done():
-
-			// context is dead, so we close the file so `CopyBuffer` won't block and fail
-			file.Close() // nolint: errcheck
-		case <-copyBufferDone:
-
-			// `CopyBuffer` doesn't block anymore, we can stop the goroutine
-			return
-		}
-	}()
-
 	// get a buffer from the pool - so we can share buffers across goroutines
 	buf := s.logCollectionBufferPool.Get()
 	defer s.logCollectionBufferPool.Put(buf)
 
-	// copy the stream into the file using the buffer, which allows us to control the size read from the file.
-	// this is blocking until there is something to read
-	numBytesWritten, err := io.CopyBuffer(file, streamReader, buf)
+	// read from the stream into the buffer
+	// this is non-blocking, it will return immediately if there is nothing to read
+	numBytesRead, err := stream.Read(buf)
 
-	// signal goroutine to exit
-	close(copyBufferDone)
+	if numBytesRead > 0 {
+
+		// write to file
+		if _, err := logFile.Write(buf[:numBytesRead]); err != nil {
+			s.Logger.WarnWithCtx(ctx, "Failed to write pod log to file",
+				"err", err.Error(),
+				"runUID", runUID)
+			return true, errors.Wrap(err, "Failed to write pod log to file")
+		}
+	}
 
 	// if error is EOF, the pod is done streaming logs (deleted/completed/failed)
 	if err == io.EOF {
 		s.Logger.DebugWithCtx(ctx, "Pod logs are done streaming", "runUID", runUID)
 		return false, nil
 	}
 
-	// nothing read, continue
-	if numBytesWritten == 0 {
-		return true, nil
-	}
+	if numBytesRead == 0 {
 
-	// if error is not EOF, log it and continue
-	if err != nil {
-		s.Logger.WarnWithCtx(ctx, "Failed to read pod log",
-			"err", err.Error(),
-			"runUID", runUID)
-		return true, errors.Wrap(err, "Failed to read pod logs")
+		// if error is not EOF, log it and continue
+		if err != nil {
+			s.Logger.WarnWithCtx(ctx, "Failed to read pod log",
+				"err", err.Error(),
+				"runUID", runUID)
+			return false, errors.Wrap(err, "Failed to read pod logs")
+		}
+
+		// nothing happened, continue
+		return true, nil
 	}
```

Comment:
You are already returning `true, nil` in the `//sanity` after this if. So you can consolidate these 2 into:
```
if err != nil && numBytesRead == 0  {
	s.Logger.WarnWithCtx(ctx, "Failed to read pod log",
 		"err", err.Error(),
 		"runUID", runUID)
 	return false, errors.Wrap(err, "Failed to read pod logs")
}